### DISCUSSION TITLE
ModuleNotFoundError: No module named 'netbox.urls'

### DISCUSSION THREAD

**Discussion by Kori-San** [netbox-community/netbox#9579] [Q&A] [answered]
Hi !

I was asked to upgrade the Netbox Service for my company from v.2.7.6 to a 3.X version.

I've been following all the steps showed in the [upgrade guide](https://docs.netbox.dev/en/stable/installation/upgrading/) all the services are up (both netbox and netbox-rq) but when I try to connect to the web's UI I have an "Internal Server Error" message displaying.

I've been looking for a cause for this message for days and i think i finally find it.

In my '/var/log/supervisor/netbox-stderr-XXXXXX.log' there is only one error displaying on loop:
```During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/dist-packages/django/utils/deprecation.py", line 94, in __call__
    response = response or self.get_response(request)
  File "/usr/local/lib/python3.6/dist-packages/django/core/handlers/exception.py", line 36, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/dist-packages/django/core/handlers/exception.py", line 90, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/dist-packages/django/core/handlers/exception.py", line 128, in handle_uncaught_exception
    callback, param_dict = resolver.resolve_error_handler(500)
  File "/usr/local/lib/python3.6/dist-packages/django/urls/resolvers.py", line 597, in resolve_error_handler
    callback = getattr(self.urlconf_module, 'handler%s' % view_type, None)
  File "/usr/local/lib/python3.6/dist-packages/django/utils/functional.py", line 80, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/usr/local/lib/python3.6/dist-packages/django/urls/resolvers.py", line 577, in urlconf_module
    return import_module(self.urlconf_name)
  File "/usr/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'netbox.urls'
```
So it looks like Netbox can't find a urls.py.

What I've done and / or tried:
- I've used v2.11.12 first then I upgraded to v3.0.0 and finally to v3.2.5
- I've copy / pasted:
    - netbox/netbox/configuration.example.py -> netbox/netbox/configuration.py
    - contrib/*.service -> /lib/systemd/system/
    - contrib/gunicorn.py -> ./gunicorn.py
    - The apache config
- I tried a fresh install but the error is still here so it might be linked to the database.
- I upgrade from python3.6 to python3.8

I tried to seek for similar issues on Netbox or any other Open Source project but I didn't find anything, the only mention of an urls.py is from the Django doc.

Here are all the files matching the "\*url\*.py" glob:
```
> /opt/netbox $ find -name *url*.py
./venv/lib/python3.8/site-packages/django_rq/tests/urls.py
./venv/lib/python3.8/site-packages/django_rq/urls.py
./venv/lib/python3.8/site-packages/social_django/urls.py
./venv/lib/python3.8/site-packages/django/core/checks/urls.py
./venv/lib/python3.8/site-packages/django/contrib/flatpages/urls.py
./venv/lib/python3.8/site-packages/django/contrib/admin/templatetags/admin_urls.py
./venv/lib/python3.8/site-packages/django/contrib/staticfiles/urls.py
./venv/lib/python3.8/site-packages/django/contrib/auth/urls.py
./venv/lib/python3.8/site-packages/django/contrib/admindocs/urls.py
./venv/lib/python3.8/site-packages/graphene_django/tests/urls_inherited.py
./venv/lib/python3.8/site-packages/graphene_django/tests/urls_pretty.py
./venv/lib/python3.8/site-packages/graphene_django/tests/urls.py
./venv/lib/python3.8/site-packages/django_tables2/columns/urlcolumn.py
./venv/lib/python3.8/site-packages/tests/myapp/urls.py
./venv/lib/python3.8/site-packages/debug_toolbar/urls.py
./venv/lib/python3.8/site-packages/django_prometheus/urls.py
./venv/lib/python3.8/site-packages/rest_framework/urlpatterns.py
./venv/lib/python3.8/site-packages/rest_framework/utils/urls.py
./venv/lib/python3.8/site-packages/rest_framework/urls.py
./venv/lib/python3.8/site-packages/urllib3/util/url.py
./venv/lib/python3.8/site-packages/pip/_internal/models/direct_url.py
./venv/lib/python3.8/site-packages/pip/_internal/utils/urls.py
./venv/lib/python3.8/site-packages/pip/_internal/utils/direct_url_helpers.py
./venv/lib/python3.8/site-packages/pip/_vendor/urllib3/util/url.py
./netbox/dcim/api/urls.py
./netbox/dcim/urls.py
./netbox/extras/tests/dummy_plugin/api/urls.py
./netbox/extras/tests/dummy_plugin/urls.py
./netbox/extras/api/urls.py
./netbox/extras/plugins/urls.py
./netbox/extras/urls.py
./netbox/virtualization/api/urls.py
./netbox/virtualization/urls.py
./netbox/circuits/api/urls.py
./netbox/circuits/urls.py
./netbox/wireless/api/urls.py
./netbox/wireless/urls.py
./netbox/ipam/api/urls.py
./netbox/ipam/urls.py
./netbox/users/api/urls.py
./netbox/users/urls.py
./netbox/netbox/urls.py
./netbox/tenancy/api/urls.py
./netbox/tenancy/urls.py
```
Does anyone faced a similar issue ? Or might have a clue of why it isn't working ?

Thanks in advance.

**Comment by candlerb** [comment, marked as answer]
Your installation is badly broken, I can see at least two serious problems there:

* Netbox 3.2 does not work under python 3.6. Minimum requirement is now [python 3.8](https://github.com/netbox-community/netbox/releases/tag/v3.2.0), and from netbox 3.0 it was python 3.7.
* Netbox must be run inside a virtualenv, not using locally installed libraries under `/usr/local/lib`

It may be broken in other ways too that I can't see.

What I suggest is:
1. Install a new Netbox 3.2.5 system, with python 3.8+, following the fresh installation instructions. Test it and get it fully working. Make sure your startup scripts are correctly pointing to binaries inside the venv, as for example [here](https://github.com/netbox-community/netbox/blob/develop/contrib/netbox.service).
2. Drop the postgres database and recreate it from a pg_dump taken from previous 3.0 or 3.1 system
3. Run `./upgrade.sh` to bring the database up to date

Note that for versions older than 2.11, it's necessary to upgrade to 2.11.x first.  Once you've done that successfully, you can then use the pgdump from 2.11.x to import into 3.2.x (followed by `./upgrade.sh`)

**Reply by Kori-San** [reply]
Hi, thank you so much for this answer, I'm going to try your suggestion and see if it work on my case !
